source/extensions/listener_managers/listener_manager/fcds_impl.cc

```diff
@@ -0,0 +1,227 @@
+void FcdsApi::processAddedOrUpdatedFilterChains(const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+  std::vector<const envoy::config::listener::v3::FilterChain*>& updated_filter_chains, 
+  std::vector<const envoy::config::listener::v3::FilterChain*>& new_filter_chains) {
+  
+  std::vector<std::string> seen_filter_chains; 
+
+  for (const auto& fcds_resource : added_resources) {
+    envoy::config::listener::v3::FilterChain new_filter_chain =
+        dynamic_cast<const envoy::config::listener::v3::FilterChain&>(fcds_resource.get().resource());
+
+    // If filter chanin name is missing, reject the update
+    if (new_filter_chain.name().size() == 0) {
+      throw EnvoyException(fmt::format("missing filter chain(s) name for fcds config block = {}", fcds_cfg_name_));
+    }
+
+    // If there are two filter chains with the same name, reject the update
+    const std::string new_filter_chain_name = new_filter_chain.name();
+    if (std::find(seen_filter_chains.begin(), seen_filter_chains.end(), new_filter_chain_name)
+        != seen_filter_chains.end()) {
+      throw EnvoyException(fmt::format("duplicate filter chain = {}", new_filter_chain_name));
+    }
+
+    seen_filter_chains.push_back(new_filter_chain_name);
+
+    // Ignore the config update if the old and new configs are same
+    const auto hash = MessageUtil::hash(new_filter_chain);
+    if (hash == filter_chain_manager_.getFilterChainMessageHash(new_filter_chain_name)) {
+        ENVOY_LOG(debug, "fcds: new and old filter chain config are same. Ignoring update for fc={}",
+                new_filter_chain_name);
+        continue;
+    }
+
+    // This filter chain is to be added   
+    new_filter_chains.push_back(dynamic_cast<const envoy::config::listener::v3::FilterChain*>(
+        &(fcds_resource.get().resource())));
+
+    // If this is an update request for an existing filter chain, it will checked and marked for deletion here
+    markFilterChainForDeletion(new_filter_chain_name, updated_filter_chains);
+
+    ENVOY_LOG(debug, "fcds: adding/updating the filter chain = {}", new_filter_chain_name);
+
+  }
+}
+
+// Adds the filter chain to deletion as well as draining list
+void FcdsApi::markFilterChainForDeletion(std::string filter_chain_name,
+    std::vector<const envoy::config::listener::v3::FilterChain*>& updated_filter_chains)
+{
+      auto& filter_chains_message_by_name = filter_chain_manager_.getExistingFilterChainMessages();
+      auto& filter_chains_object_by_message = filter_chain_manager_.filterChainsByMessage();
+
+      auto iter = filter_chains_message_by_name.find(filter_chain_name);
+      if (iter != filter_chains_message_by_name.end()) {
+        updated_filter_chains.push_back(&(iter->second));
+
+        auto iter1 = filter_chains_object_by_message.find(iter->second);
+        if (iter1 != filter_chains_object_by_message.end()) {
+          filter_chain_manager_.addFcToDrainingList(iter1->second);
```

Comment:
Is this list being cleared when a failure occurs?
(Scenario:
1. receiving FCDS response, which adds an element to the list.
2. While continuing to process the message there's an exception.
3. receiving a new FCDS response, and while processing that, the draining list still holds the filter chain from the previous response.

Reply:
Rollback for exception cases is included in the latest commit.